File: scripts/audio_analysis.py
#!/usr/bin/env python3
"""Run Silero VAD on an audio or video file and output speech segments as JSON.

Usage: python3 scripts/audio_analysis.py <input_path> <output_json>

If given a video file, extracts audio to a temp 16kHz mono WAV first.
Outputs JSON with speech_segments and long_pauses (gaps > 300ms).
"""
import json
import os
import subprocess
import sys
import tempfile

SAMPLE_RATE = 16000
PAUSE_THRESHOLD = 0.3  # seconds — gaps longer than this are "long pauses"

AUDIO_EXTENSIONS = {'.wav', '.flac', '.mp3', '.aac', '.ogg', '.m4a'}

def is_audio_file(path):
    return os.path.splitext(path)[1].lower() in AUDIO_EXTENSIONS

def extract_audio(input_path, output_wav):
    """Extract audio from video to 16kHz mono WAV."""
    cmd = [
        'ffmpeg', '-hide_banner', '-loglevel', 'warning', '-y',
        '-i', input_path,
        '-ar', str(SAMPLE_RATE), '-ac', '1',
        output_wav
    ]
    subprocess.run(cmd, check=True)

def load_wav_as_tensor(wav_path):
    """Load 16kHz mono WAV as a torch tensor (avoids torchaudio/torchcodec)."""
    import torch
    from scipy.io import wavfile
    import numpy as np

    rate, data = wavfile.read(wav_path)
    assert rate == SAMPLE_RATE, f"Expected {SAMPLE_RATE}Hz, got {rate}Hz"
    # Normalize to float32 [-1, 1]
    if data.dtype == np.int16:
        data = data.astype(np.float32) / 32768.0
    elif data.dtype == np.int32:
        data = data.astype(np.float32) / 2147483648.0
    elif data.dtype == np.float32:
        pass
    else:
        data = data.astype(np.float32)
    return torch.from_numpy(data)

def run_vad(wav_path):
    """Run Silero VAD and return speech segments."""
    from silero_vad import load_silero_vad, get_speech_timestamps

    model = load_silero_vad()
    audio = load_wav_as_tensor(wav_path)
    timestamps = get_speech_timestamps(audio, model, sampling_rate=SAMPLE_RATE)

    segments = []
    for ts in timestamps:
        segments.append({
            'start': round(ts['start'] / SAMPLE_RATE, 3),
            'end': round(ts['end'] / SAMPLE_RATE, 3),
        })
    return segments

def find_long_pauses(segments):
    """Identify gaps > PAUSE_THRESHOLD between consecutive speech segments."""
    pauses = []
    for i in range(1, len(segments)):
        gap_start = segments[i - 1]['end']
        gap_end = segments[i]['start']
        duration = round(gap_end - gap_start, 3)
        if duration > PAUSE_THRESHOLD:
            pauses.append({
                'start': gap_start,
                'end': gap_end,
                'duration': duration,
            })
    return pauses

def main():
    if len(sys.argv) != 3:
        print("Usage: python3 scripts/audio_analysis.py <input_path> <output_json>", file=sys.stderr)
        sys.exit(1)

    input_path = sys.argv[1]
    output_json = sys.argv[2]

    if not os.path.exists(input_path):
        print(f"Input not found: {input_path}", file=sys.stderr)
        sys.exit(1)

    tmp_wav = None
    try:
        if is_audio_file(input_path):
            # Audio file — convert to 16kHz mono WAV for Silero
            tmp_fd, tmp_wav = tempfile.mkstemp(suffix='.wav', prefix='vad_')
            os.close(tmp_fd)
            print(f"Converting to 16kHz mono: {os.path.basename(input_path)}", file=sys.stderr)
            extract_audio(input_path, tmp_wav)
            wav_path = tmp_wav
        else:
            # Video file — extract audio
            tmp_fd, tmp_wav = tempfile.mkstemp(suffix='.wav', prefix='vad_')
            os.close(tmp_fd)
            print(f"Extracting audio from video: {os.path.basename(input_path)}", file=sys.stderr)
            extract_audio(input_path, tmp_wav)
            wav_path = tmp_wav

        print("Running Silero VAD...", file=sys.stderr)
        segments = run_vad(wav_path)
        pauses = find_long_pauses(segments)

        result = {
            'source_file': os.path.abspath(input_path),
            'sample_rate': SAMPLE_RATE,
            'speech_segments': segments,
            'long_pauses': pauses,
        }

        os.makedirs(os.path.dirname(os.path.abspath(output_json)), exist_ok=True)
        with open(output_json, 'w') as f:
            json.dump(result, f, indent=2)

        print(f"Speech segments: {len(segments)}", file=sys.stderr)
        print(f"Long pauses: {len(pauses)}", file=sys.stderr)
        print(output_json)

    finally:
        if tmp_wav and os.path.exists(tmp_wav):
            os.unlink(tmp_wav)

if __name__ == '__main__':
    main()

File: scripts/audio_analysis.rb
#!/usr/bin/env ruby
# Runs Silero VAD speech analysis on an audio or video file.
# Caches results in library.yaml if provided.
#
# Usage: ruby scripts/audio_analysis.rb <audio_or_video_path> [library.yaml]
# Output: prints the analysis JSON path to stdout.
#
# If library.yaml is provided and has a cached speech_analysis for this file,
# returns the cached path immediately without re-running.

require 'yaml'
require 'date'
require 'fileutils'
require 'shellwords'

input_path = ARGV[0]
library_yaml = ARGV[1]

abort "Usage: ruby scripts/audio_analysis.rb <audio_or_video_path> [library.yaml]" unless input_path
abort "File not found: #{input_path}" unless File.exist?(input_path)

# Check library.yaml cache
if library_yaml && File.exist?(library_yaml)
  lib = YAML.safe_load(File.read(library_yaml), permitted_classes: [Date])
  if lib && lib['videos']
    entry = lib['videos'].find { |v| v['path'] == input_path }
    # Also check sync_audio path for dual-system setups
    entry ||= lib['videos'].find { |v| v.dig('sync_audio', 'path') == input_path }

    if entry && entry['speech_analysis']
      cached = entry['speech_analysis']
      # Resolve relative to library transcripts dir
      if library_yaml && !File.absolute_path?(cached.to_s)
        lib_dir = File.dirname(library_yaml)
        cached = File.join(lib_dir, 'transcripts', cached)
      end
      if File.exist?(cached)
        $stderr.puts "Using cached speech analysis: #{cached}"
        puts cached
        exit 0
      else
        $stderr.puts "Cached path not found, re-running: #{cached}"
      end
    end
  end
end

# Determine output path
basename = File.basename(input_path, File.extname(input_path))
if library_yaml && File.exist?(library_yaml)
  output_dir = File.join(File.dirname(library_yaml), 'transcripts')
else
  output_dir = File.dirname(input_path)
end
FileUtils.mkdir_p(output_dir)
output_json = File.join(output_dir, "#{basename}_speech_analysis.json")

# Run Python script
script_dir = File.expand_path(File.dirname(__FILE__))
python_script = File.join(script_dir, 'audio_analysis.py')

$stderr.puts "Running speech analysis on #{File.basename(input_path)}..."
result = system("python3", python_script, input_path, output_json)
abort "Speech analysis failed" unless result && File.exist?(output_json)

# Cache in library.yaml if provided
if library_yaml && File.exist?(library_yaml)
  lib = YAML.safe_load(File.read(library_yaml), permitted_classes: [Date])
  if lib && lib['videos']
    # Find matching entry by video path or sync_audio path
    entry = lib['videos'].find { |v| v['path'] == input_path }
    entry ||= lib['videos'].find { |v| v.dig('sync_audio', 'path') == input_path }

    if entry
      # Store filename only (relative to transcripts dir)
      entry['speech_analysis'] = File.basename(output_json)
      lib['last_updated'] = Date.today.to_s
      File.write(library_yaml, YAML.dump(lib))
      $stderr.puts "Cached speech analysis in library.yaml"
    end
  end
end

puts output_json
